src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -433,51 +475,58 @@ std::optional<unsigned> CodegenDataModelProvider::TryFindEndpointIndex(EndpointI
     return std::make_optional<unsigned>(idx);
 }
 
-EndpointId CodegenDataModelProvider::NextEndpoint(EndpointId before)
+DataModel::EndpointEntry CodegenDataModelProvider::NextEndpoint(EndpointId before)
 {
     const uint16_t lastEndpointIndex = emberAfEndpointCount();
 
     std::optional<unsigned> before_idx = TryFindEndpointIndex(before);
     if (!before_idx.has_value())
     {
-        return kInvalidEndpointId;
+        return DataModel::EndpointEntry::kInvalid;
     }
 
     // find the first enabled index
     for (uint16_t endpoint_idx = static_cast<uint16_t>(*before_idx + 1); endpoint_idx < lastEndpointIndex; endpoint_idx++)
     {
         if (emberAfEndpointIndexIsEnabled(endpoint_idx))
         {
-            mEndpointIterationHint = endpoint_idx;
-            return emberAfEndpointFromIndex(endpoint_idx);
+            mEndpointIterationHint                 = endpoint_idx;
+            DataModel::EndpointEntry endpointEntry = DataModel::EndpointEntry::kInvalid;
+            endpointEntry.id                       = emberAfEndpointFromIndex(endpoint_idx);
+            auto endpointInfo                      = GetEndpointInfo(endpointEntry.id);
```

Comment:
Again, seems like GetEndpointInfo should just take an already-validated index....

Reply:
Added a new GetEndpointInfoAtIndex() function and called it here.